### ISSUE TITLE
Type error with example on front page

### ISSUE THREAD

**Issue by oliver-batchelor** [gelisam/category-syntax#2]
I'm using ghc 8.0.1, maybe something has changed?

I'm quite interested in the idea of using category-syntax for defining neural networks, which would seem like a natural fit - it's a toss up between building a graph with observable sharing or using a category abstraction.

NN.hs:65:15: error:
    • Couldn't match type ‘a2’ with ‘b1’
      ‘a2’ is a rigid type variable bound by
        the type signature for:
          rearrange :: forall a1 a2 b1 b2.
                       Either (Either a1 a2) (Either b1 b2)
                       -> Either (Either a1 b1) (Either a2 b2)
        at NN.hs:63:14
      ‘b1’ is a rigid type variable bound by
        the type signature for:
          rearrange :: forall a1 a2 b1 b2.
                       Either (Either a1 a2) (Either b1 b2)
                       -> Either (Either a1 b1) (Either a2 b2)
        at NN.hs:63:14
      Expected type: Either (Either a1 a2) (Either b1 b2)
                     -> Either (Either a1 b1) (Either a2 b2)
        Actual type: Either (Either a1 b1) (Either a2 b2)
                     -> Either (Either a1 b1) (Either a2 b2)
    • In the expression: (returnC)
      In an equation for ‘rearrange’: rearrange = (returnC)
    • Relevant bindings include
        rearrange :: Either (Either a1 a2) (Either b1 b2)
                     -> Either (Either a1 b1) (Either a2 b2)
          (bound at NN.hs:65:1)

**Comment by oliver-batchelor** [comment]
-ddump-splices doesn't seem to give anything useful here, is there something better to see what is being generated?

NN.hs:(67,15)-(70,4): Splicing expression
    syntax
      [| do { ((x1_a4WT, x2_a4WU), (y1_a4WV, y2_a4WW)) <- getInput;
              returnC ((x1_a4WT, y1_a4WV), (x2_a4WU, y2_a4WW)) } |]
  ======>
    returnC

**Comment by gelisam** [comment]
> Type error with example on front page

category-syntax is  still far from complete, and as such the examples in the Readme are meant to show what I want to do with the project, not the current state of the project. See [`Tests.hs`](https://github.com/gelisam/category-syntax/blob/master/src/Tests.hs) to see what works so far.

> is there something better to see what is being generated?

Yes, see how `Tests.hs` is using `showExp` and `pprintQ` to simplify the output. There is also `debugSyntax`, but that's mainly to debug the algorithm, as it displays extra meta-data along with the output.

Note that I plan to rewrite the whole thing once I figure out how to represent [free categories with structure](https://www.reddit.com/r/haskell/comments/52f77p/how_would_you_express_the_free_category_in_haskell/d7k0ou8), so I don't plan to put in more work in the old version until I do.

> for defining neural networks, which would seem like a natural fit

Hmm, would it?

My understanding of neutral networks is that they have a very large number of nodes, so you wouldn't write down a description of the graph using thousands of variables in order to give it to `$(syntax ...)`, that would be way too tedious. And even if you did, it would generate hundreds of thousands of `swap` and `associate` calls, meaning you'd probably be spending an uncomfortably large portion of your runtime routing data between nodes instead of summing inputs or calculating how much error should backpropagate.

Or maybe you were not thinking of describing the connections between the individual neurons, but rather the connection between the layers? That would make more sense, except my understanding of neutral networks is that each layer feeds directly into the next one, there is no graph at all, only a line, in which case category-syntax brings you nothing over `(>>>)` except the ability to name the intermediate states between the layers.

**Comment by gelisam** [comment]
Ah! I think I've figured out why you thought cagegory-syntax was a good fit: you want a graph, and a category is a graph with self loops and associativity constraints. But category-syntax isn't meant to help you define _categories_, it is meant to help you define _morphisms_ in an existing category, a data type you have defined and which has a Category instance.

**Comment by oliver-batchelor** [comment]
Right - you want an easy way to define a graph which is where category-syntax could work. There is such a data structure which would look something like this...  (Vastly simplified - hopefully with some static dimension types in there)

data Network a b where
   Linear :: Int -> Int -> Network Tensor Tensor
   TanH :: Network Tensor Tensor
  Plus :: Network (Tensor, Tensor) -> Tensor
  Compose :: Network a b -> Network b c -> Network a c
  ... plus all the other category plumbing (maybe it can use one of the Free categories as a base?)

The trick is to define it in terms of high level operations (e.g. things which operate over a whole image or layer of the neural network, rather than individual neurons), then we can use Cuda libraries to actually implement the transformations.

Yes as you say I could just overload (>>>) - but modern neural networks are much more complicated than a sequence of layers, they can be very hairy and as such it's very useful to be able to label intermediate states and all that plumbing can be quite unwieldy to define!

There's also lots of other interesting uses, I've thought it would be great for defining image processing DSLs as well - for example using Vulkan.

Very interested to see what you come up with - probably I'll see how far I can get with the observable sharing approach for now. Thanks for the information!

Cheers,
Oliver

**Comment by gelisam** [comment]
> modern neural networks are much more complicated than a sequence of layers, they can be very hairy and as such it's very useful to be able to label intermediate states and all that plumbing can be quite unwieldy to define!

I see! In that case, yes, category-syntax and FreeCartesianCategory would be helpful - if they were completed.

**Comment by oliver-batchelor** [comment]
Fascinating discussion, this kind of abstraction seems useful for all kinds of things. Would gladly help if I can/test things out! Keen to hear how this goes..

Observable sharing only seems to go so far before I'd add ugly hacks to make it work. e.g. if I were to want to annotate part of the syntax tree it makes it harder to refer to the 'function' as opposed to the result.
